rust/kernel/io_mem.rs

```diff
@@ -0,0 +1,195 @@
+            Ok(unsafe { concat_idents!(rust_helper_, $name)(ptr as _) })
+        }
+    };
+}
+
+macro_rules! define_write {
+    ($name:ident, $try_name:ident, $type_name:ty) => {
+        /// Writes IO data to the given offset, known at compile time.
+        ///
+        /// If the offset is not known at compile time, the build will fail.
+        pub fn $name(&self, value: $type_name, offset: usize) {
+            Self::check_offset::<$type_name>(offset);
+            let ptr = self.ptr.wrapping_add(offset);
+            // SAFETY: The type invariants guarantee that `ptr` is a valid pointer. The check above
+            // guarantees that the code won't link if `offset` makes the write go out of bounds
+            // (including the type size).
+            unsafe { concat_idents!(rust_helper_, $name)(value, ptr as _) }
+        }
+
+        /// Writes IO data to the given offset.
+        ///
+        /// It fails if/when the offset (plus the type size) is out of bounds.
+        pub fn $try_name(&self, value: $type_name, offset: usize) -> Result {
+            if !Self::offset_ok::<$type_name>(offset) {
+                return Err(Error::EINVAL);
+            }
+            let ptr = self.ptr.wrapping_add(offset);
+            // SAFETY: The type invariants guarantee that `ptr` is a valid pointer. The check above
+            // returns an error if `offset` would make the write go out of bounds (including the
+            // type size).
+            unsafe { concat_idents!(rust_helper_, $name)(value, ptr as _) };
+            Ok(())
+        }
+    };
+}
+
+impl<const SIZE: usize> IoMem<SIZE> {
+    /// Tries to create a new instance of a memory block.
+    ///
+    /// The resource described by `res` is mapped into CPU's address space so that it can be
+    /// accessed directly. It is also consumed by this function so that it can't be mapped again
+    /// to a different address.
+    pub fn try_new(res: Resource) -> Result<Self> {
+        // Check that the resource has at least `SIZE` bytes in it.
+        if res.size < SIZE.try_into()? {
+            return Err(Error::EINVAL);
+        }
+
+        // Try to map the resource.
+        // SAFETY: Just mapping the a memory range.
+        let addr = unsafe { rust_helper_ioremap(res.offset, res.size) };
+        if addr.is_null() {
+            Err(Error::ENOMEM)
+        } else {
+            // INVARIANT: `addr` is non-null and was returned by `ioremap`, so it is valid.
+            Ok(Self { ptr: addr as usize })
+        }
+    }
+
+    const fn offset_ok<T>(offset: usize) -> bool {
+        if let Some(end) = offset.checked_add(core::mem::size_of::<T>()) {
+            if end <= SIZE {
+                return true;
+            }
+        }
+        false
+    }
+
+    const fn check_offset<T>(offset: usize) {
+        if !Self::offset_ok::<T>(offset) {
+            crate::build_error!("IoMem offset overflow");
+        }
+    }
+
+    define_read!(readb, try_readb, u8);
```

Comment:
How about just name them `read_u8, read_u16, read_u32, read_u64`?

Reply:
No, the names were chosen to match those of the kernel.

Reply:
Well, it's not always best to name them exactly as C does. `try_write_u32` certainly feels a lot more Rust-y than `try_writew`.

Reply:
I just made a mistake writing `writew` rather than `writel`; this is RISC-V uses `b/h/w/d` while kernel uses `b/w/l/q` and I confused myself. One reason to prefer `u8/u16/u32/u64`!

Reply:
I'm not claiming it's _always_ best. I'm claiming that this was an explicit and conscious choice for this interface. The idea is to reduce friction when writing drivers similar to existing C ones; in case you haven't seen it yet, here's the converted PL061 GPIO driver: https://lwn.net/Articles/863459/

Reply:
> I just made a mistake writing `writew` rather than `writel`; this is because RISC-V uses `b/h/w/d` while kernel uses `b/w/l/q` and I confused myself. One reason to prefer `u8/u16/u32/u64`!

I'm sorry to hear that.

However, as discussed in the past: our target audience is kernel developers. At least for now, when we have to make a choice between being clearer for Rust or kernel developers, we'll choose kernel developers.

Reply:
Gary, this is Linux code, not Rust code. Whether the names feel "Rust-y" or not is not really important -- what current kernel developers feel is. We need to be mindful of introducing naming changes in very well-known APIs, specially if the example driver is going to be compared with the C version side-by-side.

I prefer names like `_u32` myself too, but just because I think seeing the type is simpler. While it is tempting to improve naming, it should be discussed with others when we are in mainline, not before and on our own.

Reply:
Well, my confusion comes from the fact that I work with assembly and hardware, not that I am a Rust developer. I don't think it's really a friction for kernel developers to switch to `u8/u16/u32/u64`: they could easily understand that. Using `u8/u16/u32/u64` has an additional advantage that it's less friction for  new developers and people who work on lower levels that uses a different convention for sizes. I don't really see much advantage for keep using x86-centric `bwlq` convention.